### PR TITLE
[multistage] clean up stage metadata and mailbox usage on instance/server

### PR DESCRIPTION
**Pull request by walterddr**
step 1 of #10657 
finalized advanced version of #10670 & #10671 

Details
===
- StageMetadata is split into several pieces:
    - `DispatchablePlanMetadata` --> this is the original usage of StageMetadata, during `DispatchablePlanVisitor` and gradually fill in the contextual info for dispatch and worker assignment
    - `StageMetadata` this only contains the execution related info for a stage. It contains
        - `List<WorkerMetadata>` --> a list of worker metadata related to a specific worker for this stage
- `VirtualServer` is no more
    - Replaced by `QueryServerInstance` only used by `QueryDispatcher`
    - `VirtualServerAddress` should be deprecated, only the `workerID` part of it should stay

With this change
- `QueryPlan` now serializes `WorkerMetadata` in a global ID map based on `(stageId << 16) + workerId`
    - TODO: query plan is still being dispatch on a per worker basis in`QueryDispatcher.dispatch`, it should changed to a per `QueryServerInstance` basis.
- `VirtualServer` or `ServerInstance` can no longer be used by `QueryRunner`, instead `VirtualServerAddress` is always used: we will gradually replace it with global ID instead. 
- `StageMetadata` and `WorkerMetadata` employs a generic `customProperty` field that can support future extension of stage and worker level information. instead of hard-coding POJOs related to a specific use case.

### CODE REVIEW

File: pinot-common/src/main/proto/worker.proto [hunk 1]
```diff
@@ -72,21 +72,17 @@ message StagePlan {
...
 }
 
-message StageMetadata {
-  repeated string instances = 1;
-  repeated string dataSources = 2;
-  map<string, SegmentMap> instanceToSegmentMap = 3;
+message WorkerMetadata {
+  string instances = 1;
+  string dataSources = 2;
+  map<string, SegmentList> instanceToSegment = 3;
```
Comment (walterddr) [inline pinot-common/src/main/proto/worker.proto:81]: this is no longer needed to point from instance to table type to segments, b/c worker metadata contextually only work on 1 worker

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java [hunk 2]
```diff
@@ -39,13 +44,15 @@
 public class QueryPlan {
   private final List<Pair<Integer, String>> _queryResultFields;
   private final Map<Integer, StageNode> _queryStageMap;
-  private final Map<Integer, StageMetadata> _stageMetadataMap;
+  private final Map<Integer, WorkerMetadata> _stageMetadataMap;
+  private final Map<Integer, DispatchablePlanMetadata> _dispatchablePlanMetadataMap;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java:48]: - 1st map is for distributedStagePlan. one can argue it doesn't need to be materialized until distributeStagePlan is constructed
- 2nd map is for dispatcher only - to dispatch stage to physical server.

nit: rename _stageMetdataMap to _workerMetadataMap

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerMetadata.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ *   <li>the partition mechanism of the data being execute on this worker</li>
+ * </ul>
+ *
+ * <p>WorkerMetadata should allow {@link org.apache.pinot.query.planner.stage.MailboxSendNode} and
+ * {@link org.apache.pinot.query.planner.stage.MailboxReceiveNode} to derive the appropriate set of mailboxes to be
+ * created when executing the worker.</p>
+ */
+public class WorkerMetadata {
+  private final QueryServerAddress _queryServerAddress;
+  private final TableInfo _tableInfo;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerMetadata.java:42]: additional fields will be put in later for WorkerMetadata such as `PartitionInfo`
we can potentially add more info for optimizations on server runtime

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/service/QueryServerTest.java [hunk 4]
```diff
@@ -143,10 +138,9 @@ public void testWorkerAcceptsWorkerRequestCorrect(String sql)
     }
   }
 
-  private static boolean isMetadataMapsEqual(StageMetadata left, StageMetadata right) {
-    return left.getServerInstances().equals(right.getServerInstances())
-        && left.getServerInstanceToSegmentsMap().equals(right.getServerInstanceToSegmentsMap())
-        && left.getScannedTables().equals(right.getScannedTables());
+  private boolean isMetadataMapsEqual(int stageId, DispatchablePlanMetadata stageMetadata,
+      Map<Integer, WorkerMetadata> metadataMap) {
+    return true;
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/service/QueryServerTest.java:143]: actually take that back. i can directly compare distributed plan worker metadata with query plan one instead of having to compare dispatchable plan metadata.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java [hunk 5]
```diff
@@ -290,10 +284,8 @@ private InstanceResponseBlock processServerQuery(ServerQueryRequest serverQueryR
   }
 
   private boolean isLeafStage(DistributedStagePlan distributedStagePlan) {
-    int stageId = distributedStagePlan.getStageId();
-    VirtualServer serverInstance = distributedStagePlan.getServer();
-    StageMetadata stageMetadata = distributedStagePlan.getMetadataMap().get(stageId);
-    Map<String, List<String>> segments = stageMetadata.getServerInstanceToSegmentsMap().get(serverInstance.getServer());
+    WorkerMetadata stageMetadata = distributedStagePlan.getCurrentWorkerMetadata();
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:287]: workerMetadata

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/OperatorUtils.java [hunk 6]
```diff
@@ -68,7 +68,7 @@ public static String canonicalizeFunctionName(String functionName) {
     return functionName;
   }
 
-  public static void recordTableName(OperatorStats operatorStats, StageMetadata operatorStageMetadata) {
+  public static void recordTableName(OperatorStats operatorStats, DispatchablePlanMetadata operatorStageMetadata) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/OperatorUtils.java:71]: this should be WorkerMetadata

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java [hunk 7]
```diff
@@ -45,108 +43,100 @@ private QueryPlanSerDeUtils() {
...
+    Map<Integer, Worker.WorkerMetadata> protoMap = new HashMap<>();
+    for (Map.Entry<Integer, WorkerMetadata> e : metadataMap.entrySet()) {
       protoMap.put(e.getKey(), toWorkerStageMetadata(e.getValue()));
     }
     return protoMap;
   }
 
-  private static Worker.StageMetadata toWorkerStageMetadata(StageMetadata stageMetadata) {
-    Worker.StageMetadata.Builder builder = Worker.StageMetadata.newBuilder();
+  private static Worker.WorkerMetadata toWorkerStageMetadata(WorkerMetadata workerMetadata) {
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java:119]: toProtoWorkerMetadata

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java [hunk 8]
```diff
@@ -39,13 +45,15 @@
 public class QueryPlan {
   private final List<Pair<Integer, String>> _queryResultFields;
   private final Map<Integer, StageNode> _queryStageMap;
-  private final Map<Integer, StageMetadata> _stageMetadataMap;
+  private final List<StageMetadata> _stageMetadataList;
+  private final Map<Integer, DispatchablePlanMetadata> _dispatchablePlanMetadataMap;
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java:49]: These maps can be simplified into List (key is always 0 to `numStages - 1`). Do you see the need of keeping it as map long term wise?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java [hunk 9]
```diff
@@ -237,6 +239,7 @@ protected RelRoot compileQuery(SqlNode sqlNode, PlannerContext plannerContext)
     SqlNode validated = validate(sqlNode, plannerContext);
     RelRoot relation = toRelation(validated, plannerContext);
     RelNode optimized = optimize(relation, plannerContext);
+    RelDecorrelator.decorrelateQuery(optimized, RelBuilder.create(_config));
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java:242]: Is this related?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/ExplainPlanStageVisitor.java [hunk 10]
```diff
@@ -62,7 +63,8 @@ public static String explain(QueryPlan queryPlan) {
     }
 
     // the root of a query plan always only has a single node
-    VirtualServer rootServer = queryPlan.getStageMetadataMap().get(0).getServerInstances().get(0);
+    QueryServerInstance rootServer = new ArrayList<>(
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/ExplainPlanStageVisitor.java:66]: You can do `.iterator().next()` to get the only element from a set

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/QueryServerInstance.java [hunk 11]
```diff
@@ -24,48 +24,42 @@
...
-
-  private final ServerInstance _server;
-  private final int _virtualId;
-
-  public VirtualServer(ServerInstance server, int virtualId) {
-    _server = server;
-    _virtualId = virtualId;
+public class QueryServerInstance {
+  private final String _hostname;
+  private final int _port;
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/QueryServerInstance.java:34]: What is this port? Do we need it?

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java (hunk 8 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java:49]: keeping it as a map will probably be changed going forward. currently this map is not used anywhere in wire protocol so we should be able to individually change this going forward. will follow up if we needed to change this in the batch-dispatcher support PR

File: pinot-common/src/main/proto/worker.proto [hunk 12]
```diff
@@ -70,23 +70,17 @@ message QueryResponse {
 
 message StagePlan {
   int32 stageId = 1;
-  string instanceId = 2;
+  string virtualAddress = 2;
```
Comment (walterddr) [inline pinot-common/src/main/proto/worker.proto:73]: - we will keep virtualAddress here indicating the host/port and workerID together, cuz to a stage plan this will always be the same and non-nullable. 
- for mailbox/opchain decoupling purpose we will use mailboxID as a unique identifier generate from planner going forward and will be put into StageMetadata/WorkerMetadata

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/ExplainPlanStageVisitor.java [hunk 13]
```diff
@@ -78,10 +79,10 @@ public static String explain(QueryPlan queryPlan) {
    *
    * @return a query plan associated with
    */
-  public static String explainFrom(QueryPlan queryPlan, StageNode node, VirtualServer rootServer) {
+  public static String explainFrom(QueryPlan queryPlan, StageNode node, QueryServerInstance rootServer) {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/ExplainPlanStageVisitor.java:82]: TODO. this was not tested. adding test in follow up PR

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+  public List<WorkerMetadata> getWorkerMetadataList() {
+    return _workerMetadataList;
+  }
+
+  public Map<String, String> getCustomProperties() {
+    return _customProperties;
+  }
+
+  public static class Builder {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java:47]: i am inclinde to use Builder/Extractor mechanism b/c we can always extend & modify the Builder/Extractor easily without worry of changing the underlying StageMetadata format going forward.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerMetadata.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    public WorkerMetadata build() {
+      return new WorkerMetadata(_virtualServerAddress, _customProperties);
+    }
+
+    public void putAllCustomProperties(Map<String, String> customPropertyMap) {
+      _customProperties.putAll(customPropertyMap);
+    }
+  }
+
+  public static Map<String, List<String>> getTableSegmentsMap(WorkerMetadata workerMetadata) {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerMetadata.java:94]: we can refactor this into a util going forward if we decided upon a common string ser/de mechanism for all custom properties.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java (hunk 8 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/QueryPlan.java:49]: related to: https://github.com/apache/pinot/pull/10673#discussion_r1175377455 will clean up afterwards once we introduce per-server dispatch logic

File: pinot-common/src/main/proto/worker.proto [hunk 16]
```diff
@@ -70,23 +70,17 @@ message QueryResponse {
...
 
-message SegmentMap {
-  map<string, SegmentList> tableTypeToSegmentList = 1;
-}
-
-message SegmentList {
-  repeated string segments = 1;
+message WorkerMetadata {
+  string virtualAddress = 1;
+  map<string, string> customProperty = 2;
```
Comment (kishoreg) [inline pinot-common/src/main/proto/worker.proto:85]: it will be good to split this into 2. metadata that's global.. across all stages and another per stageNode.

File: pinot-common/src/main/proto/worker.proto [hunk 17]
```diff
@@ -70,23 +70,17 @@ message QueryResponse {
...
 message StagePlan {
   int32 stageId = 1;
-  string instanceId = 2;
+  string virtualAddress = 2;
   StageNode stageRoot = 3;
-  map<int32, StageMetadata> stageMetadata = 4;
+  repeated StageMetadata stageMetadata = 4;
 }
 
 message StageMetadata {
```
Comment (kishoreg) [inline pinot-common/src/main/proto/worker.proto:78]: Also, a stage having multiple stageNode is a bit confusing.. should we call this StagePlan?

StageMetadata is really StagePlan Metadata.

Comment (walterddr) [reply pinot-common/src/main/proto/worker.proto:78]: this is a legacy issue that needs to be clean up next --> right now the stageMetadata list is including all the stages in the request. 

next follow up is to only send the stageMetadata with resolved mailbox info (right now mailbox info is resolved on server)

File: pinot-common/src/main/proto/worker.proto (hunk 16 above)
Comment (walterddr) [reply pinot-common/src/main/proto/worker.proto:85]: there's no per stage node metadata, (they will be associated with each stage Node).
the split is per-stage and per-worker
